agent/hcp/telemetry_provider.go

```diff
@@ -46,114 +48,138 @@ type hcpProviderImpl struct {
-	refreshInterval := telemetryCfg.RefreshConfig.RefreshInterval
-	// refreshInterval must be greater than 0, otherwise time.Ticker panics.
-	if refreshInterval <= 0 {
-		return nil, fmt.Errorf("invalid refresh interval: %d", refreshInterval)
-	}
-
-	cfg := &dynamicConfig{
-		Endpoint:        telemetryCfg.MetricsConfig.Endpoint,
-		Labels:          telemetryCfg.MetricsConfig.Labels,
-		Filters:         telemetryCfg.MetricsConfig.Filters,
-		RefreshInterval: refreshInterval,
+// defaultDisabledCfg disables metric collection and contains default config values.
+func defaultDisabledCfg() *dynamicConfig {
+	return &dynamicConfig{
+		labels:          map[string]string{},
+		filters:         client.DefaultMetricFilters,
+		refreshInterval: defaultTelemetryConfigRefreshInterval,
+		endpoint:        nil,
+		disabled:        true,
 	}
+}
 
-	t := &hcpProviderImpl{
-		cfg:       cfg,
+// NewHCPProvider initializes and starts a HCP Telemetry provider.
+func NewHCPProvider(ctx context.Context, hcpClient client.Client) *hcpProviderImpl {
+	h := &hcpProviderImpl{
+		// Initialize with default config values.
+		cfg:       defaultDisabledCfg(),
 		hcpClient: hcpClient,
 	}
 
-	go t.run(ctx, refreshInterval)
+	go h.run(ctx)
 
-	return t, nil
+	return h
 }
 
 // run continously checks for updates to the telemetry configuration by making a request to HCP.
-func (h *hcpProviderImpl) run(ctx context.Context, refreshInterval time.Duration) {
-	ticker := time.NewTicker(refreshInterval)
+func (h *hcpProviderImpl) run(ctx context.Context) {
+	// Try to initialize config once before starting periodic fetch.
+	h.updateConfig(ctx)
+
+	ticker := time.NewTicker(h.cfg.refreshInterval)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ticker.C:
-			if newCfg := h.getUpdate(ctx); newCfg != nil {
-				ticker.Reset(newCfg.RefreshInterval)
+			if newRefreshInterval := h.updateConfig(ctx); newRefreshInterval > 0 {
+				ticker.Reset(newRefreshInterval)
 			}
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
-// getUpdate makes a HTTP request to HCP to return a new metrics configuration
-// and updates the hcpProviderImpl.
-func (h *hcpProviderImpl) getUpdate(ctx context.Context) *dynamicConfig {
+// updateConfig makes a HTTP request to HCP to update metrics configuration held in the provider.
+func (h *hcpProviderImpl) updateConfig(ctx context.Context) time.Duration {
 	logger := hclog.FromContext(ctx).Named("telemetry_config_provider")
 
 	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
 	defer cancel()
 
 	telemetryCfg, err := h.hcpClient.FetchTelemetryConfig(ctx)
 	if err != nil {
+		// Only disable metrics on 404 or 401 to handle the case of an unlinked cluster.
+		// For other errors such as 5XX ones, we continue metrics collection, as these are potentially transient server-side errors.
+		apiErr, ok := err.(*runtime.APIError)
+		if ok && (apiErr.IsCode(404) || apiErr.IsCode(401)) {
```

Comment:
@jjti only targeting 404 and 401, but lmk if you want us to target 4XX class more generally? I don't think we need to care about the other ones but you might have more CCM context than me!

Reply:
> if you want us to target 4XX class more generally?

I'm open to other thoughts, but imo we should disable on any client-side status code response (4XX). All of them suggest the client was unable to get the agent telemetry config for some reason related to this configuration request, rather than a server side issue

Reply:
I'd vote for any 4xx codes.